core/src/main/scala/org/apache/spark/rpc/netty/Inbox.scala

```diff
@@ -0,0 +1,229 @@
+                    // If the sender asks a reply, we should send the error back to the sender
+                    context.sendFailure(e)
+                  } else {
+                    context.finish()
+                    throw e
+                  }
+              }
+
+            case OnStart => {
+              endpoint.onStart()
+              if (supportConcurrent) {
+                synchronized {
+                  enableConcurrent = true
+                }
+              }
+            }
+            case OnStop =>
+              dispatcher.removeRpcEndpointRef(endpoint)
+              endpoint.onStop()
+              assert(isEmpty, "OnStop should be the last message")
+              return true
+            case Associated(remoteAddress) =>
+              endpoint.onConnected(remoteAddress)
+            case Disassociated(remoteAddress) =>
+              endpoint.onDisconnected(remoteAddress)
+            case AssociationError(cause, remoteAddress) =>
+              endpoint.onNetworkError(cause, remoteAddress)
+          }
+        }
+
+        synchronized {
+          // "enableConcurrent" will be set to false after `onStop` is called, so we should check it
+          // every time.
+          if (!enableConcurrent && workerCount != 1) {
+            // If we are not the only one worker, exit
+            skipFinally = true
+            workerCount -= 1
+            return false
+          }
+          message = messages.poll()
+          if (message == null) {
+            skipFinally = true
+            workerCount -= 1
+            return false
+          }
+        }
+      }
+      return false
+    } finally {
+      if (!skipFinally) {
+        // Reset `workerCount` if some exception is thrown.
+        synchronized {
+          workerCount -= 1
+        }
+      }
+    }
+  }
+
+  def post(message: InboxMessage): Unit = {
+    val dropped =
+      synchronized {
+        if (stopped) {
+          // We already put "OnStop" into "messages", so we should drop further messages
+          true
+        } else {
+          messages.add(message)
+          false
+        }
+      }
+    if (dropped) {
+      onDrop()
+    }
+  }
+
+  def stop(): Unit = synchronized {
+    // The following codes should be in `synchronized` so that we can make sure "OnStop" is the last
+    // message
+    if (!stopped) {
+      // We should disable concurrent here. Then when RpcEndpoint.onStop is called, it's the only
+      // thread that is processing messages. So `RpcEndpoint.onStop` can release its resources
+      // safely.
+      enableConcurrent = false
+      stopped = true
+      messages.add(OnStop)
+      // Note: The concurrent events in messages will be processed one by one.
+    }
+  }
+
+  protected def onDrop(message: Any): Unit = {
+    logWarning(s"Drop ${message} because $endpointRef is stopped")
+  }
+
+  def isEmpty: Boolean = messages.isEmpty
+
+  protected def safelyCall(endpoint: RpcEndpoint)(action: => Unit): Unit = {
```

Comment:
Not sure this is helping with readability; I'd rather have the `try...catch` inline in the only call site (`process`).